### PR TITLE
br/stream: Retry apply when meet store errors

### LINKED ISSUE #32974: Port the retry rule from full restore (`IngestSST`) for incremental restore (`Apply`)
## Enhancement

Currently, restore would fail (or even worse, lost data) consistently because of the absent retry strategy.

We'd better port it from the full restore.

### PR DESCRIPTION
**Pull request by YuJuncen**
### What problem does this PR solve?

Issue Number: close #32974 

Problem Summary:

See the issue.

### What is changed and how it works?

1. Made a new "Backoffer" named `overRegionsInRange`, which would retry a function over all regions in some range, it would handle common errors like `EpochNotMatch` or something likewise.
2. Adapt it to the Log Restore.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
It can retry properly.
image
(The reason of many retry logs burst is probably region split and made many requests sent to stale epoch.)
- [ ] No code

### Release note

```release-note
None
```

### CODE REVIEW

File: br/pkg/restore/import_retry.go [hunk 1]
```diff
@@ -0,0 +1,196 @@
+package restore
```
Comment (3pointer) [inline br/pkg/restore/import_retry.go:1]: miss license.

File: br/pkg/restore/import_retry.go [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+		r.StoreError.GetNotLeader() != nil ||
+		r.StoreError.GetRegionNotFound() != nil {
+		return fromThisRegion
+	}
+
+	return fromStart
+}
+
+func (r *rpcResult) StrategyForRetryGoError() retryStrategy {
+	if r.Err == nil {
```
Comment (3pointer) [inline br/pkg/restore/import_retry.go:170]: these seems unreachable. It can be removed.

File: br/pkg/restore/import_retry.go [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+	if r.Err != nil {
+		return r.StrategyForRetryGoError()
+	}
+	return r.StrategyForRetryStoreError()
+}
+
+func (r *rpcResult) StrategyForRetryStoreError() retryStrategy {
+	if r.StoreError == nil {
+		return giveUp
+	}
```
Comment (3pointer) [inline br/pkg/restore/import_retry.go:158]: ditto

File: br/pkg/restore/import_retry.go [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+
+	if r.StoreError.GetServerIsBusy() != nil ||
+		r.StoreError.GetNotLeader() != nil ||
+		r.StoreError.GetRegionNotFound() != nil {
```
Comment (3pointer) [inline br/pkg/restore/import_retry.go:162]: what about epochNotMatch?

Comment (YuJuncen) [reply br/pkg/restore/import_retry.go:162]: They would be retry from rescanning the regions in range.

File: br/pkg/restore/import_retry.go (hunk 2 above)
Comment (YuJuncen) [reply br/pkg/restore/import_retry.go:170]: This would be called at `StrategyForRetry`.

File: br/pkg/restore/import_retry.go (hunk 4 above)
Comment (YuJuncen) [reply br/pkg/restore/import_retry.go:162]: (Because epoch not match may be caused by split / merge, which we cannot retry in the same region)

**Comment by YuJuncen** [comment]
Error: all the go tests should be run in serial.

**Review by 3pointer — APPROVED** [review]
LGTM

**Comment by YuJuncen** [comment]
[](https://ci.pingcap.net/blue/organizations/jenkins/tidb_ghpr_check/detail/tidb_ghpr_check/55948/#step-94-log-16)[](https://ci.pingcap.net/blue/organizations/jenkins/tidb_ghpr_check/detail/tidb_ghpr_check/55948/#step-94-log-17)[2022-03-22T05:31:36.497Z] go: golang.org/x/tools@v0.1.8 requires
[2022-03-22T05:31:36.497Z] 	github.com/yuin/goldmark@v1.4.1: missing go.sum entry; to add it:
[2022-03-22T05:31:36.497Z] 	go mod download github.com/yuin/goldmark

🤔 But no errors in local...

**Comment by YuJuncen** [comment]
(can we trigger the CI via close and reopen?)

File: br/pkg/restore/import_retry.go [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+	}
+	// For other errors, like `ServerIsBusy`, `RegionIsNotInitialized`, just trivially backoff.
+	time.Sleep(o.rs.ExponentialBackoff())
+	return true
+}
+
+// Run executes the `regionFunc` over the regions in `o.start` and `o.end`.
+// It would retry the errors according to the `rpcResponse`.
+func (o *OverRegionsInRangeController) Run(ctx context.Context, f RegionFunc) error {
+	if !o.rs.ShouldRetry() {
```
Comment (joccau) [inline br/pkg/restore/import_retry.go:106]: retryStatus works on some regions。
Retry time(16) is small for a mount of regions ?

**Review by joccau — APPROVED** [review]
LGTM

**Pull request merged**